src/target/riscv/riscv-013.c

```diff
@@ -4304,9 +4304,26 @@ static int riscv013_halt_go(struct target *target)
 		target_list_t *entry;
 		list_for_each_entry(entry, &dm->target_list, list) {
 			struct target *t = entry->target;
-			t->state = TARGET_HALTED;
-			if (t->debug_reason == DBG_REASON_NOTHALTED)
-				t->debug_reason = DBG_REASON_DBGRQ;
+			uint32_t dmstatus2 = dmstatus;
+			if (!get_field(dmstatus, DM_DMSTATUS_ALLHALTED) &&
+					!get_field(dmstatus, DM_DMSTATUS_ALLUNAVAIL)) {
+				/* Only some harts were halted. Need to check
+				 * status of each one individually. */
+				riscv013_info_t *info = get_info(t);
+				dmcontrol = set_dmcontrol_hartsel(dmcontrol, info->index);
+				if (dmi_write(target, DM_DMCONTROL, dmcontrol) != ERROR_OK)
+					return ERROR_FAIL;
+				dm->current_hartid = info->index;
+				if (dmi_read(target, &dmstatus2, DM_DMSTATUS) != ERROR_OK)
+					return ERROR_FAIL;
+			}
+			if (get_field(dmstatus2, DM_DMSTATUS_ALLHALTED)) {
+				t->state = TARGET_HALTED;
+				if (t->debug_reason == DBG_REASON_NOTHALTED)
+					t->debug_reason = DBG_REASON_DBGRQ;
+			} else if (get_field(dmstatus2, DM_DMSTATUS_ALLUNAVAIL)) {
+				t->state = TARGET_UNAVAILABLE;
+			}
 		}
 	}
 	/* The "else" case is handled in halt_go(). */
```

Comment:
It would be good to handle both cases at one place - but this can be done sometime later.